core/commands/add.go

```diff
@@ -263,112 +262,162 @@ You can now refer to the added file in a gateway, like so:
 
-		var sizeChan chan int64
-		s, found := req.Values()["size"]
-		if found {
-			sizeChan = s.(chan int64)
-		}
+				var bar *pb.ProgressBar
+				if progress {
+					bar = pb.New64(0).SetUnits(pb.U_BYTES)
+					bar.ManualUpdate = true
+					bar.ShowTimeLeft = false
+					bar.ShowPercent = false
+					bar.Output = os.Stderr
+					bar.Start()
+				}
 
-		lastFile := ""
-		lastHash := ""
-		var totalProgress, prevFiles, lastBytes int64
-
-	LOOP:
-		for {
-			select {
-			case out, ok := <-outChan:
-				if !ok {
-					if quieter {
-						fmt.Fprintln(res.Stdout(), lastHash)
-					}
-					break LOOP
+				var sizeChan chan int64
+				s, found := req.Values()["size"]
+				if found {
+					sizeChan = s.(chan int64)
 				}
-				output := out.(*coreunix.AddedObject)
-				if len(output.Hash) > 0 {
-					lastHash = output.Hash
-					if quieter {
-						continue
-					}
 
-					if progress {
-						// clear progress bar line before we print "added x" output
-						fmt.Fprintf(res.Stderr(), "\033[2K\r")
-					}
-					if quiet {
-						fmt.Fprintf(res.Stdout(), "%s\n", output.Hash)
-					} else {
-						fmt.Fprintf(res.Stdout(), "added %s %s\n", output.Hash, output.Name)
+				lastFile := ""
+				lastHash := ""
+				var totalProgress, prevFiles, lastBytes int64
+
+			LOOP:
+				for {
+					select {
+					case out, ok := <-outChan:
+						if !ok {
+							if quieter {
+								fmt.Fprintln(os.Stdout, lastHash)
+							}
+
+							break LOOP
+						}
+						output := out.(*coreunix.AddedObject)
+						if len(output.Hash) > 0 {
+							lastHash = output.Hash
+							if quieter {
+								continue
+							}
+
+							if progress {
+								// clear progress bar line before we print "added x" output
+								fmt.Fprintf(os.Stderr, "\033[2K\r")
+							}
+							if quiet {
+								fmt.Fprintf(os.Stdout, "%s\n", output.Hash)
+							} else {
+								fmt.Fprintf(os.Stdout, "added %s %s\n", output.Hash, output.Name)
+							}
+
+						} else {
+
+							if !progress {
+								continue
+							}
+
+							if len(lastFile) == 0 {
+								lastFile = output.Name
+							}
+							if output.Name != lastFile || output.Bytes < lastBytes {
+								prevFiles += lastBytes
+								lastFile = output.Name
+							}
+							lastBytes = output.Bytes
+							delta := prevFiles + lastBytes - totalProgress
+							totalProgress = bar.Add64(delta)
+						}
+
+						if progress {
+							bar.Update()
+						}
+					case size := <-sizeChan:
+						if progress {
+							bar.Total = size
+							bar.ShowPercent = true
+							bar.ShowBar = true
+							bar.ShowTimeLeft = true
+						}
+					case <-req.Context().Done():
+						re.SetError(req.Context().Err(), cmdsutil.ErrNormal)
+						return
 					}
-				} else {
-					log.Debugf("add progress: %v %v\n", output.Name, output.Bytes)
+				}
+			}
 
-					if !progress {
-						continue
-					}
+			go func() {
+				// defer order important! First close outChan, then wait for output to finish, then close re
```

Comment:
hrm... probably fine to have a warning here for now. But in the future we will probably want to make it easier on anyone writing something with a progress bar